errdefs/http_helpers.go

```diff
@@ -99,6 +101,12 @@ func FromStatusCode(err error, statusCode int) error {
 		if !IsSystem(err) && !IsUnknown(err) && !IsDataLoss(err) && !IsDeadline(err) && !IsCancelled(err) {
 			err = System(err)
 		}
+	case -1:
+		if errors.Is(err, context.Canceled) {
+			err = Cancelled(err)
+		} else {
+			err = Unknown(err)
+		}
```

Comment:
I'm a bit on the fence if this is the right location for this; if we _do_ want to include it here, perhaps it should be in the `default` section below.

That said; this function was specifically to convert a HTTP status to a typed error. IIUC, in the context cancelled cases we're handling here, it's not a "context cancelled" that's returned by the API (i.e., if a context was cancelled in the daemon code), but a "context cancelled" on the client side? 🤔

If that's indeed the case, perhaps these errors should be handled _before_ we try to convert a HTTP status returned by the daemon, e.g. in https://github.com/moby/moby/blob/46cdcd206c56172b95ba5c77b827a722dab426c5/client/request.go#L113-L116

or in https://github.com/moby/moby/blob/46cdcd206c56172b95ba5c77b827a722dab426c5/client/request.go#L121-L126

Reply:
i agree, in fact there is already a check for that in `doRequest`:

https://github.com/moby/moby/blob/46cdcd206c56172b95ba5c77b827a722dab426c5/client/request.go#L135-L139

but this does not solve the problem  

i think that a better solution will be to change: 
https://github.com/moby/moby/blob/46cdcd206c56172b95ba5c77b827a722dab426c5/client/request.go#L113-L118
to:
```go
	resp, err := cli.doRequest(ctx, req)
	switch {
	case errors.Is(err, context.Canceled), errors.Is(err, context.DeadlineExceeded):
		return serverResponse{}, err
	case err == nil:
		err = cli.checkResponseErr(resp)
	}

	return resp, errdefs.FromStatusCode(err, resp.statusCode)
```

Reply:
Thanks; yes (at least at a quick glance), I think your proposed changes look good